### PR TITLE
basic Jenkins pipeline support for notifier

### PR DESCRIPTION
**Pull request by balihb**
This is a very rough Jenkins Pipeline support with code duplication that should somehow be refactored.
An example of such pipeline is like this:
[Jenkinsfile.txt](https://github.com/uber/phabricator-jenkins-plugin/files/555993/Jenkinsfile.txt)

### CODE REVIEW

**Comment by sectioneight** [comment]
Very cool! Is there some way for me to test this? I've never used jenkinsfile/pipeline before

**Comment by balihb** [comment]
there is a repository here with the proper Jenkinsfile:
https://github.com/balihb/boot-react/blob/master/Jenkinsfile
The check step is currently failing, so I can test build failure. I'm currently using staging area for code review. I haven't modified the Phabricator build wrapper to work with Jenkins Pipeline.
I have the following build plan step in Harbormaster (make http request):
https://myci/buildByToken/buildWithParameters?job=${repository.callsign}_review&DIFF_ID=${buildable.diff}&PHID=${target.phid}&token=mytoken&STREF=${repository.staging.ref}
The job is configured as a pipeline job, with the post parameters as build parameters and the following SCM config:
jppconfig

**Comment by sectioneight** [comment]
Ok, so it sounds like this is still a Work-In-Progress commit then, I'll add a WIP label and then ping me when this is ready for review.

**Comment by aadis** [comment]
We'd be interested in this as well.

**Comment by balihb** [comment]
I'll update the PR as soon as I can.
Some notes:
- no wrapper support, just notification (might be possible, never tried)
- no coverage support since cobertura has no pipeline support. for now and the type hierarchy stops the plugin to support coverage (run vs abstractbuild. let me know if anyone knows a solution) (also refer to: https://github.com/jenkinsci/cobertura-plugin/pull/55 )
- introducing pipeline tests is a whole bunch of big work in itself, so it might take a while wile I can produce more than just some self tested code

**Comment by aadis** [comment]
As you mentioned the cobertura plugin is right now seeing active work on pipeline support. The PR#55 builds and works. 

It might be a good idea to test against that (or I can do the testing & report back if you update this PR)

**Comment by balihb** [comment]
The solution is getting more and more hacky. I can't exactly base a PR on an unreleased dependency.
So still only notification support.

**Comment by balihb** [comment]
@sectioneight if I create proper pipeline tests would this PR be eligible for merging into master? There is still no coverage and build wrapper support, but it would be great have the basic pipeline support in the released plugin.

**Comment by sectioneight** [comment]
Absolutely, if there's unit test coverage, at least a little bit of
documentation (maybe in the advanced section) and the code is reasonable
then this is purely additive. No need to implement everything all at once,
in fact we prefer smaller PRs.

Also feel free to open up a Work In Progress PR if you want to get a sanity
check / feedback on the general approach.

**Comment by sectioneight** [comment]
Oh, i see, we already have a PR :) I was assuming this was an issue while responding via email. 

Would you like us to take a look at this now or are there changes forthcoming?

**Comment by cmmata** [comment]
Is this PR still valid?

**Comment by balihb** [comment]
it is valid. I just haven't had the time to write the tests for it.

**Comment by aadis** [comment]
jenkinsci/cobertura-plugin#55 has been merged, so it'd be great if this PR can make sure cobertura support is enabled.

We're using this PR directly, and other than the coverage reporting/comparison, it's been working great!

**Comment by easyrasta** [comment]
+1

**Comment by balihb** [comment]
I finally had time to finish this up. Please try it out!
I want to follow it up with a few more PR like wrapper support and pipeline dsl support among others.

**Comment by balihb** [comment]
@sectioneight ping :D

**Comment by sectioneight** [comment]
@balihb I'm not working on this anymore, perhaps @jjx could take a look or knows somebody who's current?

**Comment by balihb** [comment]
maybe @artms as the latest commiter?

**Comment by artms** [comment]
@kageiit ideas ? I'm not that proficient with plugin... I might be able to check its' behavior on staging though...

**Comment by balihb** [comment]
anyone?

**Comment by tomposmiko** [comment]
@jjx @kageiit @artms 
I am a colleague of @balihb. We are relying on this plugin and improvement heavily. Could someone take a look at it pls and merge if possible.
Who is the authorized person?
Thanks.

**Comment by lindblandro** [comment]
I'm currently migrating our jobs to pipelines and would be very interested in using this. Has there been any progress?

**Comment by baxor** [comment]
@Pawka can you help this PR find a reviewer by chance? :)

**Comment by sascha-egerer** [comment]
Just did a test and works fine here. This is my Jenkinsfile
```
pipeline {
  agent any
  parameters {
    string(name: 'PHID', defaultValue: '', description: '')
    string(name: 'DIFF_ID', defaultValue: '', description: '')
  }
  stages {
    stage('build') {
      steps {
        sh 'echo foo'
      }
    }
  }
  post {
    always {
        script {
            // workaround. Seems to be a bug!?
            currentBuild.result = currentBuild.currentResult
        }
        step([$class: 'PhabricatorNotifier', commentOnSuccess: true, commentWithConsoleLinkOnFailure: true])
    }
  }
}
```

**Comment by kageiit** [comment]
Im ok merging this if its been staged and working without breaking backwards compatibility

**Comment by kageiit** [comment]
@balihb please sign the CLA before we can merge this

**Comment by kageiit** [comment]
@balihb Can you please also add some documentation to the advanced section

**Comment by balihb** [comment]
I've added a fix for the null pointer exception in the pipeline.
Docs is in progress.

**Comment by balihb** [comment]
it should be fixed by my latest change.

File: build.gradle [hunk 1]
```diff
@@ -47,7 +47,7 @@ dependencies {
   jenkinsPlugins 'org.jenkins-ci.plugins:credentials:1.22@jar'
 
   optionalJenkinsPlugins 'org.jenkins-ci.plugins:junit:1.6@jar'
-  optionalJenkinsPlugins 'org.jenkins-ci.plugins:cobertura:1.9.6@jar'
+  optionalJenkinsPlugins 'org.jenkins-ci.plugins:cobertura:1.11@jar'
```
Comment (kageiit) [inline build.gradle]: Why is this required? Can we decouple this change?

Comment (sascha-egerer) [reply build.gradle]: Answered in https://github.com/uber/phabricator-jenkins-plugin/pull/185#issuecomment-389227543

Comment (kageiit) [reply build.gradle]: sounds good

**Comment by balihb** [comment]
documentation added. Hope it is ok. let me know if I should change anything about it.

File: src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java [hunk 2]
```diff
@@ -85,7 +85,11 @@ public BuildResultProcessor(
     }
 
     public Result getBuildResult() {
-        return this.build.getResult();
+        if (this.build.getResult() == null) {
```
Comment (kageiit) [inline src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java]: Can you explain why we would want to return success on null? It seems a bit counterintuitive. Are there any api docs defining this new behavior?

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 3]
```diff
@@ -247,14 +257,14 @@ public final boolean perform(final AbstractBuild<?, ?> build, final Launcher lau
...
-            return false;
+            throw new AbortException();
         }
 
         resultProcessor.processRemoteComment(commentFile, commentSize);
 
         resultProcessor.sendComment(commentWithConsoleLinkOnFailure);
 
-        return true;
+        return;
```
Comment (kageiit) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: can remove this redundant return at the end of a void method

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaCoverageProvider.java [hunk 4]
```diff
@@ -114,7 +116,7 @@ private void computeCoverage() {
         }
 
         // Fallback to scanning for the reports
-        copyCoverageToJenkinsMaster(build);
+        copyCoverageToJenkinsMaster(build, workspace);
```
Comment (kageiit) [inline src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaCoverageProvider.java]: just inline `build.getWorkspace()` here and avoid the local `workspace` variable

File: src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java (hunk 2 above)
Comment (balihb) [reply src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java]: on oldschool freestyle jobs after the build reached the last build step it set the build result to success, but in pipeline there is no definite end of build blocks, so it won't explicitly set SUCCESS, hence the previous workaround, where it was needed to run this before the notifier:
```groovy
currentBuild.result = currentBuild.result ?: 'SUCCESS'
```

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaCoverageProvider.java (hunk 4 above)
Comment (balihb) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaCoverageProvider.java]: in the new Jenkins API you can't use Build, but must use Run, but the run interface won't provide a getWorkspace method. I can change this call to this:
```java
copyCoverageToJenkinsMaster(build, getWorkspace());
```
but this is the best I can do.

File: src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java (hunk 2 above)
Comment (balihb) [reply src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java]: null == success so far. IMHO it was a genuine missing null check in the plugin code. In freestyle it is always set and in pipeline it is always considered SUCCESS.

Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/BuildResultProcessor.java]: makes sense

**Comment by arcadien** [comment]
Up? everyting seems ready? Please at least allow us to download the plugin from Travis integration server, thanks

**Comment by kageiit** [comment]
You can use `./gradlew jpi` to build the plugin locally till a release is cut

**Comment by arcadien** [comment]
Hi, just done. Thanks

Le mar. 12 juin 2018 à 16:07, Gautam Korlam <notifications@github.com> a
écrit :

**Comment by bhush9** [comment]
Hi,

We in [KDE](https://www.kde.org) would very much like to use this on our build servers, is there any chance to get it released under plugin upgrades as we'd rather use it from released version instead of unstable git builds.

Thanks.
